src/grammar/ast.ts

```diff
@@ -679,4 +679,316 @@ export function __DANGER_resetNodeId() {
+            case "number":
+                return ast1.value === (ast2 as AstNumber).value;
+            case "string":
+                return ast1.value === (ast2 as AstString).value;
+            case "id":
+                return eqNames(ast1, ast2 as AstId);
+            case "method_call":
+                return (
+                    eqNames(ast1.method, (ast2 as AstMethodCall).method) &&
+                    eqExpressions(ast1.self, (ast2 as AstMethodCall).self) &&
+                    eqExpressionArrays(ast1.args, (ast2 as AstMethodCall).args)
+                );
+            case "init_of":
+                return (
+                    eqNames(ast1.contract, (ast2 as AstInitOf).contract) &&
+                    eqExpressionArrays(ast1.args, (ast2 as AstInitOf).args)
+                );
+            case "op_unary":
+                return (
+                    ast1.op === (ast2 as AstOpUnary).op &&
+                    eqExpressions(ast1.operand, (ast2 as AstOpUnary).operand)
+                );
+            case "op_binary":
+                return (
+                    ast1.op === (ast2 as AstOpBinary).op &&
+                    eqExpressions(ast1.left, (ast2 as AstOpBinary).left) &&
+                    eqExpressions(ast1.right, (ast2 as AstOpBinary).right)
+                );
+            case "conditional":
+                return (
+                    eqExpressions(
+                        ast1.condition,
+                        (ast2 as AstConditional).condition,
+                    ) &&
+                    eqExpressions(
+                        ast1.thenBranch,
+                        (ast2 as AstConditional).thenBranch,
+                    ) &&
+                    eqExpressions(
+                        ast1.elseBranch,
+                        (ast2 as AstConditional).elseBranch,
+                    )
+                );
+            case "struct_instance":
+                return (
+                    eqNames(ast1.type, (ast2 as AstStructInstance).type) &&
+                    eqParameterArrays(
+                        ast1.args,
+                        (ast2 as AstStructInstance).args,
+                    )
+                );
+            case "field_access":
+                return (
+                    eqNames(ast1.field, (ast2 as AstFieldAccess).field) &&
+                    eqExpressions(
+                        ast1.aggregate,
+                        (ast2 as AstFieldAccess).aggregate,
+                    )
+                );
+            case "static_call":
+                return (
+                    eqNames(ast1.function, (ast2 as AstStaticCall).function) &&
+                    eqExpressionArrays(ast1.args, (ast2 as AstStaticCall).args)
+                );
+        }
+    } catch (e) {
+        // In principle, the assertions "as Ast___" should not fail
+        // because ast1 and ast2 have the same kind inside the switch.
+        return false;
```

Comment:
ah, sorry, I missed you had

```
if (ast1.kind !== ast2.kind) {
        return false;
    }
```

above

`try-catch` isn't needed here, my bad